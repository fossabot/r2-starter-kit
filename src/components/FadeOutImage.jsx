/* global setTimeout */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

export default class FadeOutImage extends Component {
  static propTypes = {
    url: PropTypes.string.isRequired,
    placeholder: PropTypes.string.isRequired,
  };
  constructor(props) {
    super();
    this.state = {
      url: props.placeholder,
      blur: 5,
    };
  }
  handleImageOnload = (url) => {
    this.setState({ url, blur: 0 });
  };
  render() {
    const { url, blur } = this.state;
    const { url: selfUrl } = this.props;
    return (
      <div>
        <Image ref={(c) => (this.component = c)} url={url} blur={blur} />
        <img
          id="preload_image"
          src={selfUrl}
          onLoad={() => setTimeout(() => this.handleImageOnload(selfUrl), 500)}
          style={{ display: 'none' }}
        />
      </div>
    );
  }
}

const Image = styled.img`
  display: block;
  width: 100%;
  padding-bottom: 56.25%;
  background: ${({ url }) => `url('${url}') no-repeat center center`};
  background-size: cover;
  filter: ${({ blur }) => `blur(${blur}px)`};
  transition: 0.5s;
`;
