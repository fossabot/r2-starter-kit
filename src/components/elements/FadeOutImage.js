import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import styles from './FadeOutImage.scss';

export default class FadeOutImage extends PureComponent {
  static propTypes = {
    url: PropTypes.string.isRequired,
    placeholder: PropTypes.string.isRequired,
  };
  static defaultProps = {
    url: '',
    placeholder: '',
  };
  constructor(props) {
    super();
    this.state = {
      url: props.placeholder,
      noBlur: false,
    };
  }
  handleImageOnload = (url) => {
    this.setState({
      url,
      noBlur: true,
    });
  };
  render() {
    const { url, noBlur } = this.state;
    const { url: selfUrl } = this.props;
    return (
      <div>
        <img
          ref={(c) => (this.component = c)}
          style={{ backgroundImage: `url('${url}')` }}
          className={cx({
            [styles['placeholder']]: true,
            [styles['noBlur']]: noBlur,
          })}
        />
        <img
          src={selfUrl}
          onLoad={() => setTimeout(() => this.handleImageOnload(selfUrl), 500)}
          style={{ display: 'none' }}
        />
      </div>
    );
  }
}
