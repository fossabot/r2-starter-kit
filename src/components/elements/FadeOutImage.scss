.placeholder {
  display: block;
  width: 100%;
  padding-bottom: 56.25%;
  background-repeat: no-repeat;
  background-position: center;
  background-size: cover;
  filter: blur(5px);
  transition: 0.5s;
}

.noBlur {
  filter: blur(0px);
}
