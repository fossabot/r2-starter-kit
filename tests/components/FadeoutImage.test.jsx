/* global setTimeout */

import React from 'react';
import { mount } from 'enzyme';
import FadeOutImage from 'components/FadeOutImage';

describe('<FadeOutImage />', () => {
  const initProps = {
    url: 'https://placehold.it/800x600',
    placeholder: 'https://placehold.it/40x30',
  };

  let wrapper;
  beforeAll(() => {
    wrapper = mount(<FadeOutImage {...initProps} />);
  });

  it('should render initial component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should set selfUrl instead of placeholder after image onload', () => {
    jest.useFakeTimers();

    expect(wrapper.instance().state.url).toBe('https://placehold.it/40x30');

    wrapper.find('#preload_image').simulate('load');
    setTimeout(() => {
      expect(wrapper.instance().state.url).toBe('https://placehold.it/800x600');
    }, 500);

    jest.runAllTimers();
  });
});
