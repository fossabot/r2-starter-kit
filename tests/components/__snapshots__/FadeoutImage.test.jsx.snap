// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FadeOutImage /> should render initial component 1`] = `
<FadeOutImage
  placeholderUrl="https://placehold.it/40x30"
  url="https://placehold.it/800x600"
>
  <div>
    <img
      src="https://placehold.it/40x30"
      style={
        Object {
          "filter": "blur(5px)",
        }
      }
    />
    <img
      id="preload_image"
      onLoad={[Function]}
      src="https://placehold.it/800x600"
      style={
        Object {
          "display": "none",
        }
      }
    />
  </div>
</FadeOutImage>
`;
